File: 프로그래머스/lv2/12946. 하노이의 탑/하노이의 탑.cpp
#include <string>
#include <vector>

using namespace std;

void hanoi(vector<vector<int>>& v, int n, int from, int to, int by) {
	if (n == 1) {   // 종료 조건
        v.push_back({from, to});
    }
    else {
    	hanoi(v, n - 1, from, by, to);
        v.push_back({from, to});
        hanoi(v, n - 1, by, to, from);
    }
}

vector<vector<int>> solution(int n) {
    vector<vector<int>> answer;
    hanoi(answer, n, 1, 3, 2);
    return answer;
}

File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.py
def solution(n):
    answer = []
    hanoi(answer, n, 1, 3, 2)
    return answer

def hanoi(arr, n, src, dst, by):
    if n == 1:
        arr.append([src, dst])
    else:
        hanoi(arr, n - 1, src, by, dst)
        arr.append((src, dst))
        hanoi(arr, n - 1, by, dst, src)